### PR TITLE
Add kubectl ray cluster log command

### PR DESCRIPTION
**Pull request by chiayi**
## Why are these changes needed?

This PR adds the `kubectl ray cluster logs` command to retrieve the head node logs for a ray cluster.

Screenshot 2024-08-09 at 10 11 29 AM

Screenshot 2024-08-09 at 10 11 42 AM

## Checks

- [ ] I've made sure the tests are passing.
- Testing Strategy
   - [x] Unit tests
   - [ ] Manual tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: kubectl-plugin/pkg/cmd/cluster/log.go
Comment (andrewsykim) [inline kubectl-plugin/pkg/cmd/cluster/log.go:120]: I don't think it's enough to just download the stdout of the Ray head. We'll also want to save logs in `/tmp/ray/session_latest/logs` too. What do you think @kevin85421 @MortalHappiness?

Comment (chiayi) [reply kubectl-plugin/pkg/cmd/cluster/log.go:120]: Hmmm that makes sense. If thats the case, then the `out-dir` should be a requirement for the command then? Or do we still want the ability to print out all the logs under `/tmp/ray/session_latest/logs`

Comment (andrewsykim) [reply kubectl-plugin/pkg/cmd/cluster/log.go:120]: out-dir being required makes sense to me. I don't think we need to print out all the logs, saving to out-dir as the default makes sense!

Comment (MortalHappiness) [reply kubectl-plugin/pkg/cmd/cluster/log.go:120]: I also think both the log of the Ray head and the Ray workers should be downloaded. If displaying to stdout is needed, we can print some additional information like what [bat](https://github.com/sharkdp/bat) does:

```
$ kubectl ray cluster logs my-cluster

------------
raycluster-sample-head-wnsmd
------------
(the log of ray head)

------------
raycluster-sample-worker-small-group-9j7h7
------------
(the log of ray worker)
```

```
$ kubectl ray cluster logs my-cluster --out-dir ./log-dir
$ ls ./log-dir

raycluster-sample-head-wnsmd.log raycluster-sample-worker-small-group-9j7h7.log
```

@andrewsykim Is what I mentioned above the result you originally intended to implement?

Comment (andrewsykim) [reply kubectl-plugin/pkg/cmd/cluster/log.go:120]: Make sense, but maybe each head / worker pod should be a directory, and inside each directory is all the files in `/tmp/ray/session_latest/logs`

Comment (andrewsykim) [reply kubectl-plugin/pkg/cmd/cluster/log.go:120]: I'm okay scoping this PR to only the head pod since it should be easy to extend it later to also donwload logs for the worker pods

Comment (kevin85421) [reply kubectl-plugin/pkg/cmd/cluster/log.go:120]: > I don't think it's enough to just download the stdout of the Ray head. We'll also want to save logs in /tmp/ray/session_latest/logs too.

+1, we propose this function because Ray is not very K8s-friendly. Multiple processes are running in a single container and writing logs to files under /tmp/ray/session_latest instead of STDOUT/STDERR.

File: kubectl-plugin/README.md
Comment (MortalHappiness) [inline kubectl-plugin/README.md:76]: I think we don't need to put help messages in README.md. Users can always get the help message with the `--help` flag. Putting help messages in README may cause out of sync between the documentation and the real implementation.

File: kubectl-plugin/README.md
Comment (MortalHappiness) [inline kubectl-plugin/README.md:77]: This line is redundant.

File: kubectl-plugin/pkg/cmd/cluster/log.go
Comment (MortalHappiness) [inline kubectl-plugin/pkg/cmd/cluster/log.go:48]: It should be "log (RAY_CLUSTER_NAME) (--out-dir directory)". Required parameters should be enclosed with `()`.

File: kubectl-plugin/pkg/cmd/cluster/log.go
Comment (MortalHappiness) [inline kubectl-plugin/pkg/cmd/cluster/log.go:88]: You forget to handle the case that `--out-dir` is not provided by the user. It should be handled in the `Complete` function above.

**Review by MortalHappiness — CHANGES_REQUESTED** [review]

**Comment by andrewsykim** [comment]
@chiayi can you share the output of running the `kubectl ray cluster logs` command?

File: kubectl-plugin/pkg/cmd/cluster/log.go
Comment (chiayi) [reply kubectl-plugin/pkg/cmd/cluster/log.go:88]: If the `--out-dir` flag is not set by the user, it will fail the proceeding validation giving a `no such file or directory` error. I don't think thats a good error msg for the user so I will be adding another check that returns something along the lines of "No output directory, please set --out-dir flag". As for handling this case in the `Complete` function, I think we should keep all the validation should be kept in the `Validation` function. WDYT?

File: kubectl-plugin/pkg/cmd/cluster/log.go
Comment (chiayi) [reply kubectl-plugin/pkg/cmd/cluster/log.go:48]: Thanks! I completely flipped them. I want the `--out-dir` to be required and the ray cluster to be optional.

File: kubectl-plugin/pkg/cmd/cluster/log.go
Comment (MortalHappiness) [reply kubectl-plugin/pkg/cmd/cluster/log.go:88]: I borrow the idea from kubectl source code.

https://github.com/kubernetes/kubectl/blob/b2a7d2626dc56b2edf952ab4031fcc20a7f441e4/pkg/cmd/logs/logs.go#L233-L236

Here the Pod name is required and it is handled in the `Complete` function instead of `Validate` function.

image

**Comment by chiayi** [comment]
Screenshot 2024-09-03 at 11 01 43 AM

This is what it looks like when it downloads the file. Please let me know if I should add more logs.

Result: 
Screenshot 2024-09-03 at 11 37 01 AM

**Comment by andrewsykim** [comment]
@chiayi in your last screenshot, why is there two directories named `<pod-name>` and `<pod-name>-logs`?

**Comment by chiayi** [comment]
One of the them is the pod logs, and another one is the directory that contains all the files from `/tmp/ray/session_latest/logs/`.

**Comment by chiayi** [comment]
I'm going to rename the pod log one to include `pod-logs`. Let me know if there is something else I should as well.

**Comment by andrewsykim** [comment]
I suggest having just one directory per pod named after the pod name and just add a `stdout.log` file in there

**Comment by chiayi** [comment]
New output directory format:

Screenshot 2024-09-03 at 2 38 50 PM

File: kubectl-plugin/README.md
Comment (andrewsykim) [inline kubectl-plugin/README.md:50]: Can we reference an example that uses multiple nodes? 1 head and 2 workers is probably fine

File: kubectl-plugin/pkg/cmd/cluster/log.go
Comment (andrewsykim) [inline kubectl-plugin/pkg/cmd/cluster/log.go:168]: make this a constant

File: kubectl-plugin/pkg/cmd/cluster/log.go
Comment (andrewsykim) [inline kubectl-plugin/pkg/cmd/cluster/log.go:195]: I think this comment can be removed, the function name is self explanatory

File: kubectl-plugin/pkg/cmd/cluster/log.go
Comment (andrewsykim) [inline kubectl-plugin/pkg/cmd/cluster/log.go:184]: Remove this comment

File: kubectl-plugin/README.md
Comment (andrewsykim) [reply kubectl-plugin/README.md:50]: Ah, I forgot that we're scoping this PR to head pod only. 

Now that you implemented logs for head pod, do you think it would be a significant change to also extract logs for worker pods?

Comment (andrewsykim) [reply kubectl-plugin/README.md:50]: I'm thinking we can add an additional flag `--node-type=all|head|worker` to control what logs are extracted

Comment (andrewsykim) [reply kubectl-plugin/README.md:50]: (in the future a --node-name flag to extract only logs for a specific node would be useful too)

File: kubectl-plugin/pkg/cmd/cluster/log.go
Comment (chiayi) [reply kubectl-plugin/pkg/cmd/cluster/log.go:92]: As discussed offline, --out-dir flag will not be required and will default to current directory with the head name being the destination directory name. 

With that, I think it make sense to do the validation check in the `Validation` method. It seems like the check in the kubectl source code is checking for completion? WDYT?

File: kubectl-plugin/README.md
Comment (chiayi) [reply kubectl-plugin/README.md:51]: Added the flag and added check to make sure that currently only `head` is supported.

**Comment by chiayi** [comment]
Current flow:
Screenshot 2024-09-06 at 2 43 09 PM
Screenshot 2024-09-06 at 2 43 49 PM

**Comment by andrewsykim** [comment]
> Out-dir is no longer required and will download to current dir if not specified.

Instead of defaulting to current-dir, can we default to `./<cluster-name>` ?

**Comment by chiayi** [comment]
Screenshot 2024-09-09 at 12 01 39 PM

**Comment by andrewsykim** [comment]
FYI, the lint failure is fixed in https://github.com/ray-project/kuberay/pull/2366

**Review by andrewsykim — APPROVED** [review]
LGTM

**Comment by kevin85421** [comment]
@chiayi would you mind fixing the linter issue? Thanks!

**Comment by chiayi** [comment]
@kevin85421 I merged in the pervious linter fix, the current linter issues are being fixed in #2368.

**Review by kevin85421 — APPROVED** [review]
I will merge the PR after CI passes because Andrew has already approved the PR.

**Pull request merged**
